libcache/cache.c

```diff
@@ -0,0 +1,460 @@
+	else {
+		if ((*lhsLine)->tag < (*rhsLine)->tag) {
+			ret = -1;
+		}
+		else if ((*lhsLine)->tag == (*rhsLine)->tag) {
+			ret = 0;
+		}
+		else {
+			ret = 1;
+		}
+	}
+
+	return ret;
+}
+
+
+static ssize_t cache_writeToSet(cache_writeCb_t writeCb, cacheset_t *set, cacheline_t *line, const uint64_t offset, size_t count, int policy)
+{
+	int i, j;
+	ssize_t written = 0;
+	cacheline_t *temp = NULL, **linePtr = NULL;
+
+	if (set->count < LIBCACHE_NUM_WAYS) {
+		for (i = 0; i < LIBCACHE_NUM_WAYS; ++i) {
+			if (!IS_VALID(set->lines[i].flags)) {
+				set->lines[i] = *line;
+				break;
+			}
+		}
+
+		temp = &(set->lines[i]);
+
+		for (j = 0; j < LIBCACHE_NUM_WAYS; ++j) {
+			if (set->tags[j] == NULL) {
+				set->tags[j] = temp;
+				break;
+			}
+		}
+
+		LIST_ADD(&set->timestamps, temp);
+		set->count++;
+	}
+	else {
+		temp = set->timestamps;
+		LIST_REMOVE(&set->timestamps, temp);
+
+		/* write-back */
+		if (IS_DIRTY(temp->flags)) {
+			written = (*writeCb)(offset, temp->data, sizeof(*temp->data), count);
+			free(temp->data);
+		}
+
+		for (i = 0; i < LIBCACHE_NUM_WAYS; ++i) {
+			if (IS_VALID(set->lines[i].flags) && (set->lines[i].tag == temp->tag)) {
+				set->lines[i] = *line;
+				break;
+			}
+		}
+
+		LIST_ADD(&set->timestamps, temp);
+	}
+
+	switch (policy) {
+		case 0:
+			/* write-back */
+			SET_DIRTY(set->lines[i].flags);
+			written = count;
+			break;
+		case 1:
+			/* write-through */
+			written = (*writeCb)(offset, temp->data, sizeof(*temp->data), count);
+			break;
+		default:
+			break;
+	}
+
+	qsort(set->tags, LIBCACHE_NUM_WAYS, sizeof(cacheline_t *), cache_compareTags);
+
+	return written;
+}
+
+
+static void *cache_readFromSet(cacheset_t *set, uint64_t tag, uint64_t offset)
+{
+	unsigned char *buf = NULL;
+	cacheline_t key, *keyPtr = &key, **linePtr = NULL;
+
+	key.tag = tag;
+
+	linePtr = bsearch(&keyPtr, set->tags, LIBCACHE_NUM_WAYS, sizeof(cacheline_t *), cache_compareTags);
+
+	if (linePtr != NULL) {
+		buf = (unsigned char *)(*linePtr)->data + offset;
+
+		LIST_REMOVE(&set->timestamps, *linePtr);
+		LIST_ADD(&set->timestamps, *linePtr);
+	}
+
+	return buf;
+}
+
+
+static uint64_t cache_compOffset(const cachectx_t *cache, const uint64_t addr)
+{
+	return addr & cache->offMask;
+}
+
+
+static uint64_t cache_compSet(const cachectx_t *cache, const uint64_t addr)
+{
+	return (addr >> cache->offBitNum) & cache->setMask;
+}
+
+
+static uint64_t cache_compTag(const cachectx_t *cache, const uint64_t addr)
+{
+	return (addr >> (cache->offBitNum + cache->setBitsNum)) & cache->tagMask;
+}
+
+
+ssize_t cache_write(cachectx_t *cache, const uint64_t addr, void *buffer, size_t count, int policy)
+{
+	ssize_t wrt = 0;
+	void *dataPtr = NULL;
+	unsigned char *buf = NULL, *dest = NULL, flags = 0;
+	uint64_t index = 0, offset = 0, tag = 0;
+	cacheline_t line;
+
+	index = cache_compSet(cache, addr);
+	offset = cache_compOffset(cache, addr);
+	tag = cache_compTag(cache, addr);
+
+	dataPtr = malloc(sizeof(unsigned char) * cache->lineSize);
```

Comment:
You don't check the value returned by `malloc()`.

Reply:
Thank you for pointing that out. I'll remember that when adding proper write functionality.